# Исследование надежности заемщиков.

<div style="border:solid blue 2px; padding: 20px">
Привет Руслан! Меня зовут Андрей Петрушин, я буду проверять твой проект. Предлагаю общаться на «ты», это упрощает коммуникацию :) Но если это некомфортно - можем перейти на “Вы”.

Вижу, что ты много работал над проектом и он выполнен на хорошем уровне. Моя основная цель при проверке работы — разобраться в ошибках, если они есть и возможно, поделиться своим опытом решения подобных задач.

Я буду делать комментарии четырьмя цветами:

<font color='blue'>Комментарий ревьювера синий текст - просто текст комментария</font>
    
<font color="green">Комментарий ревьювера    
✔️ зеленый текст - все отлично, хорошее решение</font>
    
<font color="orange">Комментарий ревьвюера    
⚠️ оранжевый текст - есть рекомендации, в другой раз можно сделать иначе</font>
    
<font color="red">Комментарий ревьювера    
❌ красный текст - есть недочеты, их нужно исправить и проект будет принят</font>
    
Пожалуйста, не удаляй мои комментарии, нам обоим будет проще разобраться в недочетах, что и как исправлено. Свои комментарии тоже отмечай, например другим цветом или надписью “Комментарий студента”. 

\<font color='purple'> Этот код позволяет менять цвет текста. \</font>
</div>


# Задачи исследования:

# Осмотр данных

- посмотрим начало и конец датасета
- сделаем предварительные выводы
- поищем аномалии

# Обработка данных

- найти и обработать пропуски
- заменить некорректные типы данных
- удалить дубликаты

# Ответы на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?
- Есть ли зависимость между семейным положением и возвратом кредита в срок?
- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
- Как разные цели кредита влияют на его возврат в срок?

<font color="orange">Комментарий ревьвюера   
⚠️ Хорошо, что ты описал план работы. Но еще очень желательно дать описание задачи и данных - очень важная часть любого проекта. С них нужно обязательно начинать проект. Это поможет и тебе лучше разобраться в задании и будущему заказчику понять, что ды сделал. Можно просто скопировать сюда текст из задания, а можно дать какие-то пояснения.</font>

## Обзор данных

In [1]:
import pandas as pd

In [2]:
# чтение файла с данными и сохранение в df
df = pd.read_csv('/datasets/data.csv')

In [3]:
# вывожу первые 10 строк таблицы df
df.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

In [4]:
# получение общей информации о данных в таблице df
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


В столбцах "days_employed" и "total_income" вижу пропущенные значения.
В столбце "days_employed" множество отрицательных значений, явная ошибка.
В "education" разный регистр слов, который тоже надо исправить. Все остальные данные выглядят корректными.

<font color="green">Комментарий ревьювера    
✔️ Хорошо, ты вывел основные данные и сделал первые выводы</font>

## Предобработка данных

### Заполнение пропусков

In [5]:
# подсчёт пропусков
df.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [6]:
# подсчёт доли пропусков (посчитал одно значение, тк кол-во пропусков и значений в других столбцах одинаковое)
share = (df['days_employed'].isna().sum() / len(df['days_employed'])) * 100
share

10.099883855981417

In [7]:
# убираю минусы в столбце "days_employed"
df['days_employed'] = abs(df['days_employed'])
df.head()

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

In [8]:
sheet = df[df['days_employed'].isna()]
sheet = sheet[sheet['total_income'].isna()]
sheet

children  days_employed  dob_years education  education_id  \
12            0            NaN         65   среднее             1   
26            0            NaN         41   среднее             1   
29            0            NaN         63   среднее             1   
41            0            NaN         50   среднее             1   
55            0            NaN         54   среднее             1   
...         ...            ...        ...       ...           ...   
21489         2            NaN         47   Среднее             1   
21495         1            NaN         50   среднее             1   
21497         0            NaN         48    ВЫСШЕЕ             0   
21502         1            NaN         42   среднее             1   
21510         2            NaN         28   среднее             1   

               family_status  family_status_id gender  income_type  debt  \
12          гражданский брак                 1      M    пенсионер     0   
26           женат / замужем                 0      M  госслужащий     0   
29     Не женат / не замужем                 4      F    пенсионер     0   
41           женат / замужем                 0      F  госслужащий     0   
55          гражданский брак                 1      F    пенсионер     1   
...                      ...               ...    ...          ...   ...   
21489        женат / замужем                 0      M    компаньон     0   
21495       гражданский брак                 1      F    сотрудник     0   
21497        женат / замужем                 0      F    компаньон     0   
21502        женат / замужем                 0      F    сотрудник     0   
21510        женат / замужем                 0      F    сотрудник     0   

       total_income                           purpose  
12              NaN                   сыграть свадьбу  
26              NaN                       образование  
29              NaN  строительство жилой недвижимости  
41              NaN  сделка с подержанным автомобилем  
55              NaN                   сыграть свадьбу  
...             ...                               ...  
21489           NaN              сделка с автомобилем  
21495           NaN                           свадьба  
21497           NaN        строительство недвижимости  
21502           NaN  строительство жилой недвижимости  
21510           NaN           приобретение автомобиля  

[2174 rows x 12 columns]

Интересен факт, что в строках с пропусками пропущены значения сразу в двух столбацх.
Не могу однозначно назвать причину, возможно люди целенаправленно не указывали эту информацию.

In [9]:
# заполняю пропуски в столбцах "days_employed" и "total_income" медианными значениями
df['days_employed'] = df['days_employed'].fillna(df['days_employed'].median())
df['total_income'] = df['total_income'].fillna(df['total_income'].median())
df.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

Считаю, что заполнение пропусков медианным значением в данном случае является более корректным решением, тк среднее значение слишком смещается из-за "аномалий".

<font color="green">Комментарий ревьювера   
✔️ Все верно, согласен</font>

### Проверка данных на аномалии и исправления.

In [10]:
# проверяю столбец 'children' на аномальные значения
df['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

Чем больше детей, тем меньше строк, выглядит логично. Строчки со значениями "20" и "-1" похожи на какую-то ошибку. Предлагаю удалить их из анализа, тк их не так много.

<font color="orange">Комментарий ревьвюера   
⚠️ Анализ верный, а решение спорное. Да, можно удалить. Но мне кажется более вероятно, что перед 1 просто поставили тире, а 20 - это вероятно 2. Но это только предположения, так что твою версию принимаем.</font>

In [11]:
df = df[df['children'] <= 5]
df = df[df['children'] >= 0]
df['children'].value_counts()

0    14149
1     4818
2     2055
3      330
4       41
5        9
Name: children, dtype: int64

In [12]:
# проверяю столбец 'days_employed' на аномальные значения
# предварительно переведу значения в годовое выражение для удобства
df['days_employed'] = df['days_employed'] / 360
df.head(10)

children  days_employed  dob_years education  education_id  \
0         1      23.437981         42    высшее             0   
1         1      11.180010         36   среднее             1   
2         0      15.620618         33   Среднее             1   
3         3      11.457631         32   среднее             1   
4         0     945.183533         53   среднее             1   
5         0       2.572738         27    высшее             0   
6         0       7.997783         43    высшее             0   
7         0       0.424388         50   СРЕДНЕЕ             1   
8         2      19.249626         35    ВЫСШЕЕ             0   
9         0       6.079879         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

Вижу значение 945 лет, что не совсем похоже на реальный трудовой стаж. Предположу, что максимальный трудовой стаж равен 35-40 лет. Все превышающие значения заменю на медианное по столбцу.

<font color="red">Комментарий ревьювера    
❌ Так делать нельзя. Любое предложение по замене нужно обосновать цифрами, статистикой или логикой. У нас есть на работе сотрудники в возрасте и 70 и 85 лет. А это трудовой стаж больше 60 лет. </font>

In [13]:
df[df['days_employed'] >= 40].head()

children  days_employed  dob_years education  education_id  \
4          0     945.183533         53   среднее             1   
18         0    1111.892047         53   среднее             1   
24         1     940.422091         57   среднее             1   
25         0    1009.856915         67   среднее             1   
30         1     932.171301         62   среднее             1   

            family_status  family_status_id gender income_type  debt  \
4        гражданский брак                 1      F   пенсионер     0   
18         вдовец / вдова                 2      F   пенсионер     0   
24  Не женат / не замужем                 4      F   пенсионер     0   
25        женат / замужем                 0      M   пенсионер     0   
30        женат / замужем                 0      F   пенсионер     0   

     total_income                                purpose  
4   158616.077870                        сыграть свадьбу  
18   56823.777243     на покупку подержанного автомобиля  
24  290547.235997  операции с коммерческой недвижимостью  
25   55112.757732                   покупка недвижимости  
30  171456.067993  операции с коммерческой недвижимостью

In [14]:
df.loc[df['days_employed'] >= 40, 'days_employed'] = df['days_employed'].median()
df.sort_values(by = 'days_employed')

children  days_employed  dob_years education  education_id  \
17437         1       0.067060         31   среднее             1   
8336          0       0.067335         32    высшее             0   
6157          2       0.083876         47   среднее             1   
9683          0       0.093113         43   среднее             1   
2127          1       0.096392         31    высшее             0   
...         ...            ...        ...       ...           ...   
20909         0      38.824464         58    высшее             0   
2938          0      39.031118         65   СРЕДНЕЕ             1   
11072         0      39.177398         56   среднее             1   
8735          0      39.558146         53   Среднее             1   
16626         0      39.948738         63   среднее             1   

               family_status  family_status_id gender  income_type  debt  \
17437        женат / замужем                 0      F    сотрудник     1   
8336   Не женат / не замужем                 4      M    сотрудник     0   
6157        гражданский брак                 1      M    компаньон     0   
9683   Не женат / не замужем                 4      M    сотрудник     1   
2127         женат / замужем                 0      F    компаньон     0   
...                      ...               ...    ...          ...   ...   
20909  Не женат / не замужем                 4      F    сотрудник     0   
2938         женат / замужем                 0      F  госслужащий     0   
11072        женат / замужем                 0      M    сотрудник     0   
8735         женат / замужем                 0      M    компаньон     0   
16626        женат / замужем                 0      M    сотрудник     0   

        total_income                                purpose  
17437  166952.415427                     высшее образование  
8336   124115.373655  получение дополнительного образования  
6157   231461.185606                                свадьба  
9683   128555.897209                приобретение автомобиля  
2127    90557.994311                  получение образования  
...              ...                                    ...  
20909  112894.440598     на покупку подержанного автомобиля  
2938   258879.580616                покупка жилья для семьи  
11072  147530.985003  операции с коммерческой недвижимостью  
8735   174484.333490                покупка жилья для семьи  
16626  288665.441515                приобретение автомобиля  

[21402 rows x 12 columns]

In [15]:
# проверяю 'dob_years' на аномальные значения:
df['dob_years'].value_counts()

35    614
40    603
41    603
34    597
38    595
42    592
33    577
39    572
31    556
36    553
29    543
44    543
48    536
30    536
37    531
43    510
50    509
32    506
49    505
28    501
45    494
27    490
52    483
56    482
47    480
54    476
46    469
58    461
53    457
57    457
51    446
59    441
55    441
26    406
60    376
25    356
61    353
62    351
63    268
24    263
64    263
23    252
65    194
66    183
22    183
67    167
21    110
0     100
68     99
69     83
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

In [16]:
df[df['dob_years'] < 10]

children  days_employed  dob_years education  education_id  \
99            0       6.095057          0   Среднее             1   
149           0       7.400759          0   среднее             1   
270           3       5.201842          0   среднее             1   
578           0       6.095057          0   среднее             1   
1040          0       3.216749          0    высшее             0   
...         ...            ...        ...       ...           ...   
19829         0       6.095057          0   среднее             1   
20462         0       6.095057          0   среднее             1   
20577         0       6.095057          0   среднее             1   
21179         2       0.302686          0    высшее             0   
21313         0       3.523577          0   среднее             1   

               family_status  family_status_id gender income_type  debt  \
99           женат / замужем                 0      F   пенсионер     0   
149                в разводе                 3      F   сотрудник     0   
270          женат / замужем                 0      F   сотрудник     0   
578          женат / замужем                 0      F   пенсионер     0   
1040               в разводе                 3      F   компаньон     0   
...                      ...               ...    ...         ...   ...   
19829        женат / замужем                 0      F   сотрудник     0   
20462        женат / замужем                 0      F   пенсионер     0   
20577  Не женат / не замужем                 4      F   пенсионер     0   
21179        женат / замужем                 0      M   компаньон     0   
21313  Не женат / не замужем                 4      M   сотрудник     0   

        total_income                                 purpose  
99      71291.522491                              автомобиль  
149     70176.435951                       операции с жильем  
270    102166.458894                            ремонт жилью  
578     97620.687042  строительство собственной недвижимости  
1040   303994.134987                         свой автомобиль  
...              ...                                     ...  
19829  145017.937533                                   жилье  
20462  259193.920299                    покупка своего жилья  
20577  129788.762899                            недвижимость  
21179  240702.007382        строительство жилой недвижимости  
21313  152419.184851                       операции с жильем  

[100 rows x 12 columns]

В столбце есть нулевые значения, заменим их на среднее арифметическое - в случае возраста такой метод допустим.

<font color="orange">Комментарий ревьвюера   
⚠️ Значений достаточно много. Можно поменять, как ты предложил. Но можно и вообще не менять, поскольку обоснованных методов замены найти не представляется возможным, а этот столбец не участвует в требуемых исследованиях.</font>

In [17]:
df.loc[df['dob_years'] == 0, 'dob_years'] = df['dob_years'].mean()
df['dob_years'].value_counts()

35.000000    614
41.000000    603
40.000000    603
34.000000    597
38.000000    595
42.000000    592
33.000000    577
39.000000    572
31.000000    556
36.000000    553
44.000000    543
29.000000    543
30.000000    536
48.000000    536
37.000000    531
43.000000    510
50.000000    509
32.000000    506
49.000000    505
28.000000    501
45.000000    494
27.000000    490
52.000000    483
56.000000    482
47.000000    480
54.000000    476
46.000000    469
58.000000    461
53.000000    457
57.000000    457
51.000000    446
55.000000    441
59.000000    441
26.000000    406
60.000000    376
25.000000    356
61.000000    353
62.000000    351
63.000000    268
64.000000    263
24.000000    263
23.000000    252
65.000000    194
22.000000    183
66.000000    183
67.000000    167
21.000000    110
43.300206    100
68.000000     99
69.000000     83
70.000000     65
71.000000     58
20.000000     51
72.000000     33
19.000000     14
73.000000      8
74.000000      6
75.000000      1
Name: dob_year

In [18]:
# проверяю 'education' на аномальные значения:
# перед этим приведу всё к нижнему регистру
df['education'] = df['education'].str.lower()
df['education'].value_counts()

среднее                15136
высшее                  5237
неоконченное высшее      741
начальное                282
ученая степень             6
Name: education, dtype: int64

In [19]:
# проверяю 'education_id' на аномальные значения:
df['education_id'].value_counts()

1    15136
0     5237
2      741
3      282
4        6
Name: education_id, dtype: int64

In [20]:
# проверяю 'family_status' на аномальные значения:
df['family_status'] = df['family_status'].str.lower()
df['family_status'].value_counts()

женат / замужем          12302
гражданский брак          4160
не женат / не замужем     2799
в разводе                 1189
вдовец / вдова             952
Name: family_status, dtype: int64

In [21]:
# проверяю 'family_status_id' на аномальные значения:
df['family_status_id'].value_counts()

0    12302
1     4160
4     2799
3     1189
2      952
Name: family_status_id, dtype: int64

In [22]:
# проверяю 'gender' на аномальные значения:
df['gender'].value_counts()

F      14154
M       7247
XNA        1
Name: gender, dtype: int64

In [23]:
# вижу странное значение:
df[df['gender'] == 'XNA']

children  days_employed  dob_years            education  education_id  \
10701         0       6.551668       24.0  неоконченное высшее             2   

          family_status  family_status_id gender income_type  debt  \
10701  гражданский брак                 1    XNA   компаньон     0   

        total_income               purpose  
10701  203905.157261  покупка недвижимости

In [24]:
# проверяю 'income_type' на аномальные значения:
df['income_type'].value_counts()

сотрудник          11050
компаньон           5054
пенсионер           3839
госслужащий         1453
безработный            2
предприниматель        2
в декрете              1
студент                1
Name: income_type, dtype: int64

In [25]:
# проверяю 'debt' на аномальные значения:
df['debt'].value_counts()

0    19670
1     1732
Name: debt, dtype: int64

In [26]:
# проверяю 'total_income' на аномальные значения
# предварительно меняя тип данных на целочисленный
df['total_income'] = df['total_income'].astype('int')
df['total_income'].value_counts()

145017    2163
104693       3
145603       3
133122       3
150684       3
          ... 
150195       1
66485        1
72630        1
627639       1
126820       1
Name: total_income, Length: 18502, dtype: int64

In [27]:
# проверяю 'purpose' на аномальные значения
df['purpose'].value_counts()

свадьба                                   796
на проведение свадьбы                     772
сыграть свадьбу                           769
операции с недвижимостью                  673
покупка коммерческой недвижимости         661
покупка жилья для сдачи                   651
операции с жильем                         648
операции с коммерческой недвижимостью     646
жилье                                     642
покупка жилья                             641
покупка жилья для семьи                   640
недвижимость                              632
строительство собственной недвижимости    628
операции со своей недвижимостью           626
строительство жилой недвижимости          622
строительство недвижимости                620
покупка недвижимости                      619
покупка своего жилья                      619
ремонт жилью                              609
покупка жилой недвижимости                603
на покупку своего автомобиля              504
заняться высшим образованием      

Вроде все ок)

<font color="red">Комментарий ревьювера    
❌ В целом все правильно ,хорошо провреил значение на аномальность.
    
Но почему ты оставил непонятный пол  'XNA' ?</font>

###  Удаление дубликатов

In [28]:
# кол-во строк дубликатов
df.duplicated().sum()

71

In [29]:
# удаление этих строк
df = df.drop_duplicates().reset_index(drop=True) 
df.head()

children  days_employed  dob_years education  education_id  \
0         1      23.437981       42.0    высшее             0   
1         1      11.180010       36.0   среднее             1   
2         0      15.620618       33.0   среднее             1   
3         3      11.457631       32.0   среднее             1   
4         0       6.095057       53.0   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

<font color="green">Комментарий ревьювера   
✔️ Хорошо.</font>

### Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.

In [30]:
#новый датафрейм 'education_type'
education_type = df[['education', 'education_id']]
education_type = education_type.drop_duplicates().reset_index(drop=True)
education_type

education  education_id
0               высшее             0
1              среднее             1
2  неоконченное высшее             2
3            начальное             3
4       ученая степень             4

In [31]:
#новый датафрейм 'family_status_type'
family_status_type = df[['family_status', 'family_status_id']]
family_status_type = family_status_type.drop_duplicates().reset_index(drop=True)
family_status_type

family_status  family_status_id
0        женат / замужем                 0
1       гражданский брак                 1
2         вдовец / вдова                 2
3              в разводе                 3
4  не женат / не замужем                 4

In [32]:
#удаление из исходного датафрейма колонок "education" и "family_status"
df = df.drop(['education'], inplace=False, axis=1)
df = df.drop(['family_status'], inplace=False, axis=1)
df

children  days_employed  dob_years  education_id  family_status_id  \
0             1      23.437981       42.0             0                 0   
1             1      11.180010       36.0             1                 0   
2             0      15.620618       33.0             1                 0   
3             3      11.457631       32.0             1                 0   
4             0       6.095057       53.0             1                 1   
...         ...            ...        ...           ...               ...   
21326         1      12.581435       43.0             1                 1   
21327         0       6.095057       67.0             1                 0   
21328         1       5.870408       38.0             1                 1   
21329         3       8.645783       38.0             1                 0   
21330         2       5.512521       40.0             1                 0   

      gender income_type  debt  total_income                       purpose  
0          F   сотрудник     0        253875                 покупка жилья  
1          F   сотрудник     0        112080       приобретение автомобиля  
2          M   сотрудник     0        145885                 покупка жилья  
3          M   сотрудник     0        267628    дополнительное образование  
4          F   пенсионер     0        158616               сыграть свадьбу  
...      ...         ...   ...           ...                           ...  
21326      F   компаньон     0        224791             операции с жильем  
21327      F   пенсионер     0        155999          сделка с автомобилем  
21328      M   сотрудник     1         89672                  недвижимость  
21329      M   сотрудник     1        244093  на покупку своего автомобиля  
21330      F   сотрудник     0         82047         на покупку автомобиля  

[21331 rows x 10 columns]

<font color="green">Комментарий ревьювера   
✔️ Все хорошо.</font>

### Категоризация дохода.

In [33]:
#создам функцию категоризации доходов
def income_group(income):
    if 30000 >= income:
        return 'E'
    if 50000 >= income >= 30001:
        return 'D'
    if 200000 >= income >= 50001:
        return 'C'
    if 1000000 >= income >= 200001:
        return 'B'
    return 'A'
    

In [34]:
#создаю новый столбец с категориями
df['total_income_category'] = df['total_income'].apply(income_group)
df

children  days_employed  dob_years  education_id  family_status_id  \
0             1      23.437981       42.0             0                 0   
1             1      11.180010       36.0             1                 0   
2             0      15.620618       33.0             1                 0   
3             3      11.457631       32.0             1                 0   
4             0       6.095057       53.0             1                 1   
...         ...            ...        ...           ...               ...   
21326         1      12.581435       43.0             1                 1   
21327         0       6.095057       67.0             1                 0   
21328         1       5.870408       38.0             1                 1   
21329         3       8.645783       38.0             1                 0   
21330         2       5.512521       40.0             1                 0   

      gender income_type  debt  total_income                       purpose  \
0          F   сотрудник     0        253875                 покупка жилья   
1          F   сотрудник     0        112080       приобретение автомобиля   
2          M   сотрудник     0        145885                 покупка жилья   
3          M   сотрудник     0        267628    дополнительное образование   
4          F   пенсионер     0        158616               сыграть свадьбу   
...      ...         ...   ...           ...                           ...   
21326      F   компаньон     0        224791             операции с жильем   
21327      F   пенсионер     0        155999          сделка с автомобилем   
21328      M   сотрудник     1         89672                  недвижимость   
21329      M   сотрудник     1        244093  на покупку своего автомобиля   
21330      F   сотрудник     0         82047         на покупку автомобиля   

      total_income_category  
0                         B  
1                         C  
2                         C  
3                         B  
4                         C  
...                     ...  
21326                     B  
21327                     C  
21328                     C  
21329                     B  
21330                     C  

[21331 rows x 11 columns]

<font color="green">Комментарий ревьювера   
✔️ Все верно</font>

### Категоризация целей кредита.

In [35]:
#проверим уникальные значения в столбце "purpose"
df['purpose'].value_counts()

свадьба                                   790
на проведение свадьбы                     763
сыграть свадьбу                           760
операции с недвижимостью                  672
покупка коммерческой недвижимости         658
покупка жилья для сдачи                   649
операции с жильем                         647
операции с коммерческой недвижимостью     645
жилье                                     641
покупка жилья                             640
покупка жилья для семьи                   637
недвижимость                              631
строительство собственной недвижимости    628
операции со своей недвижимостью           623
строительство жилой недвижимости          620
покупка своего жилья                      619
строительство недвижимости                619
покупка недвижимости                      616
ремонт жилью                              604
покупка жилой недвижимости                602
на покупку своего автомобиля              504
заняться высшим образованием      

In [36]:
#напишем функцию по категоризации целей
def purpose_group(purpose):
    if 'образован' in purpose:
        return 'получение образования'
    if 'авто' in purpose:
        return 'операции с автомобилем'
    if 'свадь' in purpose:
        return 'проведение свадьбы'
    return 'операции с недвижимостью'

In [37]:
#формируем новый столбец 'purpose_category'
df['purpose_category'] = df['purpose'].apply(purpose_group)
df

children  days_employed  dob_years  education_id  family_status_id  \
0             1      23.437981       42.0             0                 0   
1             1      11.180010       36.0             1                 0   
2             0      15.620618       33.0             1                 0   
3             3      11.457631       32.0             1                 0   
4             0       6.095057       53.0             1                 1   
...         ...            ...        ...           ...               ...   
21326         1      12.581435       43.0             1                 1   
21327         0       6.095057       67.0             1                 0   
21328         1       5.870408       38.0             1                 1   
21329         3       8.645783       38.0             1                 0   
21330         2       5.512521       40.0             1                 0   

      gender income_type  debt  total_income                       purpose  \
0          F   сотрудник     0        253875                 покупка жилья   
1          F   сотрудник     0        112080       приобретение автомобиля   
2          M   сотрудник     0        145885                 покупка жилья   
3          M   сотрудник     0        267628    дополнительное образование   
4          F   пенсионер     0        158616               сыграть свадьбу   
...      ...         ...   ...           ...                           ...   
21326      F   компаньон     0        224791             операции с жильем   
21327      F   пенсионер     0        155999          сделка с автомобилем   
21328      M   сотрудник     1         89672                  недвижимость   
21329      M   сотрудник     1        244093  на покупку своего автомобиля   
21330      F   сотрудник     0         82047         на покупку автомобиля   

      total_income_category          purpose_category  
0                         B  операции с недвижимостью  
1                         C    операции с автомобилем  
2                         C  операции с недвижимостью  
3                         B     получение образования  
4                         C        проведение свадьбы  
...                     ...                       ...  
21326                     B  операции с недвижимостью  
21327                     C    операции с автомобилем  
21328                     C  операции с недвижимостью  
21329                     B    операции с автомобилем  
21330                     C    операции с автомобилем  

[21331 rows x 12 columns]

<font color="green">Комментарий ревьювера   
✔️ Все верно!</font>

## Ответы на вопросы.

### Вопрос 1: Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [38]:
children_pivot = df.pivot_table(index = ['children'], \
                                columns = ['debt'], \
                                values = 'purpose', aggfunc='count')

children_pivot['ratio'] = children_pivot[1] / children_pivot[0] * 100
children_pivot

debt            0       1      ratio
children                            
0         13028.0  1063.0   8.159349
1          4364.0   444.0  10.174152
2          1858.0   194.0  10.441335
3           303.0    27.0   8.910891
4            37.0     4.0  10.810811
5             9.0     NaN        NaN

<font color="orange">Комментарий ревьвюера   
⚠️ Если в aggfunc() добавить 'mean', то можно избавиться от второй строчки :)</font>

**Вывод: люди с детьми или без отдают кредиты примерно на одном и том же уровне**

### Вопрос 2: Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [39]:
family_status_pivot = df.pivot_table(index=['family_status_id'], columns=['debt'], values='total_income', aggfunc='count')
family_status_pivot['ratio'] = family_status_pivot[1] / family_status_pivot[0] * 100
family_status_pivot['family_status'] = family_status_type['family_status']
family_status_pivot

debt                  0    1      ratio          family_status
family_status_id                                              
0                 11334  927   8.178931        женат / замужем
1                  3749  385  10.269405       гражданский брак
2                   888   63   7.094595         вдовец / вдова
3                  1105   84   7.601810              в разводе
4                  2523  273  10.820452  не женат / не замужем

**Семейное положение влияет**

Самые низкие показатели у разведенных и вдов — можно предположить, что возвращаемость кредитов зависит в том числе от наличия кормильца или супруга в семье.

### Вопрос 3: Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [40]:
monthly_income_pivot = df.pivot_table(index=['total_income_category'], columns=['debt'], values='education_id', aggfunc='count')
monthly_income_pivot['ratio'] = monthly_income_pivot[1] / monthly_income_pivot[0] * 100
monthly_income_pivot

debt                       0     1      ratio
total_income_category                        
A                         23     2   8.695652
B                       4659   354   7.598197
C                      14569  1353   9.286842
D                        328    21   6.402439
E                         20     2  10.000000

**Есть незначительная разница в возвращаемости кредитов от уровня доходов**

Хуже всего покатель у людей с низкими доходами, что не удивительно.

### Вопрос 4: Как разные цели кредита влияют на его возврат в срок?

In [41]:
credit_purpose_pivot = df.pivot_table(index=['purpose_category'], columns=['debt'], values='education_id', aggfunc='count')
credit_purpose_pivot['ratio'] = credit_purpose_pivot[1] / credit_purpose_pivot[0] * 100
credit_purpose_pivot

debt                         0    1      ratio
purpose_category                              
операции с автомобилем    3879  400  10.311936
операции с недвижимостью  9971  780   7.822686
получение образования     3619  369  10.196187
проведение свадьбы        2130  183   8.591549

В результате группирования по категории цели кредита видим, что количество должников в категориях "автомобиль" и "образование" выделяется в большую сторону.

<font color="green">Комментарий ревьювера   
✔️ Все верно, хорошие выводы</font>

## Общий вывод:

Есть зависимости между целями кредита, семейным положением, уровнем дохода и количеством детей в семье. Нельзя сказать, что они кардинально отличаются, но 2% на больших числах могут оказать существенное влияние на кредитный портфель банка. Особое внимание следует уделить вдовам/вдовцам или молодоженам — с ними могут риски по невозврату кредитов выше.

<font color="orange">Комментарий ревьвюера   
⚠️ Еще можно посмотреть на более сложные зависмости. Например совместить цель кредита и доход или семейное положение. Разница между группами будет более заметная. Это не сложно сделать с помощью группировки по двум столбцам </font>     
    
       
<font color='blue'>Молодец, очень хороший проект, выполнены все пункты задания. Замечательно, что пишешь выводы после каждого раздела. Есть небольшие замечания, но это ничего страшного, их исправление позволит нам вместе сделать проект лучше и хорошо разобраться в теме. Работа с пропусками, дубликатами не так проста, и это очень пригодится тебе в дальнейшем!</font>